### DISCUSSION TITLE
-

### DISCUSSION THREAD

**Discussion by spinets** [vim/vim#12736] [General]
R.I.P

**Comment by ghost** [comment]
There are two others in the Vim organization, they probably have commit access

**Reply by LunarWatcher** [reply]
They do: https://groups.google.com/g/vim_dev/c/dq9Wu5jqVTw

**Comment by nyngwang** [comment]
Now who can manifest the quality of Vim?

**Comment by ebits21** [comment]
I sincerely hope the 'benevolent dictator' model is revised and vim re-organizes its development.

RIP Bram.

**Reply by nyngwang** [reply]
"[...] 'benevolent dictator' model " is unnecessary, unless you aren't one of the maintainers.

**Comment by FHU-yezi** [comment]
If he already set the successor feature in GitHub, someone he chosed can move this repo to he/she's account or make it public archived.

About this: https://docs.github.com/en/enterprise-cloud@latest/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-access-to-your-personal-repositories/maintaining-ownership-continuity-of-your-personal-accounts-repositories

**Comment by oozeRyze** [comment]
> Given the [tragic news](https://groups.google.com/g/vim_announce/c/tWahca9zkt4) about @brammool, is there anyone here with the commit access to the repository?

This is not a problem that is difficult to solve.The rest of the people in the organization can contact Github support to grant access and permissions to teams and members.

R.I.P. Bram

**Comment by yesudeep** [comment]
RIP Bram.

**Comment by xaizek** [comment]
Hope Vim's future will be no less impactful on IT sphere than its past.  Vim has some great co-maintainers, so I don't think it's too much to expect.

**Comment by wbpayne22902** [comment]
I imagine that the two others who are in the vim organization will step up and things will be more or less the same after a period of grieving over Bram's loss.  I was so sad to hear of Bram's passing myself after having used Vim myself for years, he will be sorely missed, that is certain.  RIP Bram, I'm sure you are programming something amazing in the afterlife.

**Comment by tonymec** [comment]
> Given the [tragic news](https://groups.google.com/g/vim_announce/c/tWahca9zkt4) about @brammool, is there anyone here with the commit access to the repository?

I don't know, but it can certainly be arranged; and in any case, if nothing else avails, the "master" repository can be moved to http://hg.256bit.org/vim i.e. the Mercurial clone maintained by @chrisbra which is already up to patchlevel 9.0.1677. The next patch exists, but only AFAIK as an email version: it can be brought in without too much hassle. For the rest, a solution shall be found, $DEITY willing.

**Comment by mikl** [comment]
I’m sure it’ll work out, but it might take a little time for everything to get settled. Have a little patience.

**Comment by Dawei-Simayi** [comment]
And is there something that we should consider? With the older generation contributors getting older, should we let the younger generation contributors participate in maintaining the open source repo?

**Comment by Shane-XB-Qian** [comment]
this is NOT the good timing to discuss this, please show your respect!

**Reply by Computer2340** [reply]
At least for me, this was a thought I had too, not because I selfishly want new Vim features to keep coming down the pipe, but because I want @brammool's baby to keep receiving love and attention. I'm sure there are people who are only concerned about the product but not the person, but unless stated otherwise, I'm going to assume people thinking about the governance of Vim going forward are interested in continuing Bram's legacy.

**Comment by 0Delta** [comment]
Big thanks to @k-takata @chrisbra and Dr. Chip for actively considering maintaining the repository.

We need to debate whether admins will continue to commit, whether community consensus will continue to commit, or if there is a better way.

**Comment by peacock0803sz** [comment]
We have to wait for @chrisbra to return from vacation....

**Comment by dertuxmalwieder** [comment]
I’ll continue to make Windows builds, it’s the least I can do.

**Comment by skywind3000** [comment]
I am astonished that Bram was still updating vim three weeks ago while suffering bad health situation. He sees Vim as his own child, and he surely hopes that Vim will continue to thrive after his death.

Does he have any last wishes or plans regarding how Vim should continue to develop? Did he leave some words to the community?
This should be an important thing to figure out, does anyone have connection with Bram's family?

Of course, they are busy preparing for the funeral. It would be appreciated if they could provide some information about this.

In addition, I am glad to see there are so many people ready to offer help but don't know how. The community needs leadership right now, fortunately there are some people still have access to this repository, I hope they can take on this great responsibility.

One last thing, I wish to donate for whoever can maintain Vim on a monthly basis, but I don't know how to do it.
I believe many people in this community would like to make a donation because we want Vim keep thrive just like Bram did.

**Comment by yegappan** [comment]
Bram maintained the list of planned Vim enhancements, fixes and future plans in the [todo.txt file.](https://github.com/vim/vim/blob/master/runtime/doc/todo.txt)

Bram gave importance to fixing crashes, security issues, filetype updates and fixes for existing features over adding new features.  I think the first priority should be to incorporate the currently [outstanding PRs](https://github.com/vim/vim/pulls) and fix [issues ](https://github.com/vim/vim/issues) reported against existing features.

He planned for a lot more enhancements to the Vim9 script (e.g. support for types, enums, etc.).  It is unfortunate that this is going to take quite a bit of effort for others to implement.

**Reply by chrisbra** [reply]
Hi,
I haven't still thought it through, as I am still shocked by the bad news (and I am also a bit afraid of the repsonsibility to maintain Vim and given the high quality Bram has provided over the years)

I am currently thinking to work through the existing PRs and mainly merge any un-contraditionary changes (mostly bug-fixes) and for now I'd like to continue using Brams working method, e.g. each change a new patch number (possibly with the exception of adding minor patches for runtime file updates as well to be able to find changes better).

But this quite likely won't scale well, because we need to increment the minor version with each patch, so we cannot just click the merge button. Once this is done, I'd like to release a vim9.1 or vim 10 quite soon, after which I'd like to move on to a bit more modern working with git and github (not sure about the minor versioning numbers).

Also, given the commitment of Bram, I am afraid we won't be able to keep up with the speed that Bram has provided. I'll probably add a few more people to the Vim Organization who are known for there contribution to Vim as well. But even then I don't think we can spend 7 days/week on working with Vim. So it will most-likely be a more traditional way to add changes. Don't expect big changes for now (like TreeSitter Integration, finishing up the Vim9 Class Definition, etc), it will most likely be Bug-Fixes, and Security Fixes (huntr bugs, etc).

Also there are quite a few areas, which would need some work, and I am not sure we have the knowledge to do so :( 
- better/fixed Key-handling in Windows
- renovating the Gui with support for Wayland and/or GTK4
- better integration with the console vim on Windows using the new Windows Terminal
- fixing the terminal options handling (which is quiet complicated)
- infrastructure work (the Vim homepage still has a lot of problems with the Database). Not to mention, the design could be improved a bit. and I am not sure we should stay with osdn.net given the recent problems...

And finally, I don't know yet how to handle ICCF donations, this needs to be worked out with Brams family (if Bram has mentioned how he would like to have this worked out).

So plenty of things to do. These are my rough ideas. Feel free to comment.

**Reply by jessepav** [reply]
> He planned for a lot more enhancements to the Vim9 script (e.g. support for types, enums, etc.). It is unfortunate that this is going to take quite a bit of effort for others to implement.

I do agree -- I'm not sure if anyone will grasp Bram's vision and implementation of vim9script in the way he did, and it is a huge undertaking to get up to speed with that aspect of Vim 9.

One of the main "Why Vim 9?" points was to "deprioritize [language] interfaces", but given that, for the time being, the development of vim9script faces significant hurdles, perhaps we should encourage the improvement of the language interfaces once again.

Browsing the histories of `if_py` and `if_lua` source files, I see that most improvements already have come from third-party contributors; and for someone generally unfamiliar with the Vim codebase, those interface sources provide a stable gateway of sorts for Vim's internals.

**Reply by mattn** [reply]
Agreed. And we will then have to discuss whether we should add features according to the sponsors' voting priorities.

**Reply by savchenko** [reply]
> renovating the Gui with support for Wayland and/or GTK4

How about "outsourcing" this to one of the many GUI clients? Similar to what nVim does.

> infrastructure work (the Vim homepage still has a lot of problems with the Database). Not to mention, the design could be improved a bit. and I am not sure we should stay with osdn.net given the recent problems...

This can be handled by the community. I'd be keen to help for example, maybe let's start a separate discussion?

> And finally, I don't know yet how to handle ICCF donations, this needs to be worked out with Brams family (if Bram has mentioned how he would like to have this worked out).

Might be the hardest part... Won't comment here, sounds like something that will be driven by people close to him / those with the knowledge about the fund and its activities.

**Reply by lilydjwg** [reply]
> Similar to what nVim does.

nVim does it very badly: limited GUI capabilities, broken support for clipboards, lack of input method support, etc. There are quite a few nVim GUIs, but none of them is free from these issues. And why do we split force into multiple competing projects (only to create software broken in different ways)?

It's a good idea to split GUI to a dedicate process, but please don't follow nVim's practice.

**Reply by h-east** [reply]
Below are the major ones included of my patch in this year.

[patch 9.0.1569](https://github.com/vim/vim/commit/2bd6a09691fc71974ca7ab5973f8e08fdd2a889a): cannot use "this.member" in lambda in class method
[patch 9.0.1466](https://github.com/vim/vim/commit/61378a15425553cec1ec15f590e70ca50c7c34a8): cannot use an object member name as a method argument
[patch 9.0.1463](https://github.com/vim/vim/commit/4c42c7eef43ff0f58fa574f7a900c8a3313f372e): virtual text truncation only works with Unicode 'encoding'
[patch 9.0.1373](https://github.com/vim/vim/commit/194555c001f2b8576483ef34511450b6e9b5e3fd): wrong text displayed when using both 'linebreak' and 'list'
[patch 9.0.1164](https://github.com/vim/vim/commit/01c5f2addf86b8d8940d71c92f6219f14a6df99e): evaluating string expression advances function line

I did bugfixes such as Vim9 script and text property.
Of course, I don't have a complete grasp of Vim, and I can't support on full-time.
But I think I can help the community a little.:handshake:

**Reply by k-takata** [reply]
> for now I'd like to continue using Brams working method, e.g. each change a new patch number

I'm thinking whether we can use the Squash and merge option of the merge button. (Somehow the patch number must be added to each PR, though.)
For https://github.com/vim/vim/pull/12740, I think we can safely use the merge button with the Squash and merge option (or with the Rebase and merge option).

@chrisbra Do you have the write access of the FTP server?
Are you going to upload each patch file to FTP? (I'm not sure if it is still needed.)

**Reply by chrisbra** [reply]
unfortunately not yet. I don't know, if anybody uses the FTP server for patches, but I'd still like to get access, so that we can keep at least the runtime files up to date (not sure if it is needed). Checking if this is possible.

**Reply by ychin** [reply]
> But this quite likely won't scale well, because we need to increment the minor version with each patch, so we cannot just click the merge button. Once this is done, I'd like to release a vim9.1 or vim 10 quite soon, after which I'd like to move on to a bit more modern working with git and github (not sure about the minor versioning numbers).

If we get rid of minor versioning numbers that increments after every commit, I think it should come with a comprehensive change to the release model. There are downstream effects with how different channels that Vim is released to the public (e.g. the Windows gVim binary builds, or MacVim). For MacVim for example, I just manually merge from upstream Vim routinely and use the current Vim's major/minor version number to identify the Vim version. If there's no more minor version number it would be hard for me to distinguish one version from another when I merge from upstream, unless there are frequent explicit minor releases (instead of say an official release every 1-2 years which is too slow). FWIW, the current model is somewhat tricky for MacVim anyway, as I usually have to wait a couple days to make sure there are no breaking changes in Vim before I make a MacVim release, as Vim does not have a proper idea of "release builds" that's well-tested so one version could be perfect while the next one could have some breaking bug. Neovim for example has explicit release cycles and they do code freeze before official releases. Not saying we have to do that but just something to think about.

>> How about "outsourcing" this to one of the many GUI clients? Similar to what nVim does.
>
> nVim does it very badly: limited GUI capabilities, broken support for clipboards, lack of input method support, etc. There are quite a few nVim GUIs, but none of them is free from these issues. And why do we split force into multiple competing projects (only to create software broken in different ways)?
>
> It's a good idea to split GUI to a dedicate process, but please don't follow nVim's practice.

I have quite a number of loose thoughts (that I need to organize better) on Vim GUIs, and would like to be involved if there are larger refactoring on the GUI interface of Vim, but because I would like to contribute and I have a vested interest in it. (I'm the maintainer of MacVim, the unofficial GUI implementation of Vim on macOS)

I do agree that the way Neovim handles GUI is not perfect, but there are some lessons we can learn from it, e.g. using a separate process, and providing more flexibility / structured information to the UI instead of just treating it like a smart terminal. There are quite a few features that I'm struggling to implement because of the current limitations on the GUI interface (e.g. pixel-smooth scrolling, overlay scroll bars, accessibility features).

> Agreed. And we will then have to discuss whether we should add features according to the sponsors' voting priorities.

Personally I think this can easily goes down a tricky path because sponsors are usually going to be a vocal minority, and are usually not Vim developers. I'm not sure how we can balance that with a more coherent vision that the dev team may have. Maybe for prioritization, sponsor voting could work, but I don't think "whether a feature should be added or not" should be up to sponsors to decide. The actual pros and (especially) cons of a feature are usually nuanced. For example, a lot of recent Vim breakages come from adding features that seem innocuous at first glance but add tons of hidden complexities due to interactions with other features. Are sponsors going to fully understand every single feature interactions and how it would be affected by a "simple" feature? While Bram's methods was somewhat old school, one thing he was good at was saying no (including to my PRs) when he felt that the change would be a net negative.

---

Other than this, I'm also shocked at the recent news… I'm willing to help out more depending on what help is needed.

**Reply by milanglacier** [reply]
We can also create a dedicated branch and keep the branch forever

**Comment by Peter-developer01** [comment]
R.I.P. Bram Moolenaar! :heart:

**Comment by vim-ml** [comment]
Hi,

Op zo 6 aug. 2023 18:17 schreef Christian Brabandt <
***@***.***>:

> Hi,
> I haven't still thought it through, as I am still shocked by the bad news
>

You're not alone there...

I am currently thinking to work through the existing PRs and mainly merge
> any un-contraditionary changes (mostly bug-fixes)
>

I could also read through them to see what they are all about, since you're
still on your holiday. Shall I add a LGTM comment for the PRs I've seen?

Christ van Willegen

**Comment by skywind3000** [comment]
As @yegappan and @chrisbra mentioned above: 

> first priority should be to incorporate the currently [outstanding PRs](https://github.com/vim/vim/pulls) and fix [issues ](https://github.com/vim/vim/issues)reported

> I am currently thinking to work through the existing PRs and mainly merge any un-contraditionary changes

I like the conservative and consistent development approach of Vim, which is also prudent when it comes to new features.
This makes Vim more robust, and I will not be afraid of frequently breaking my vimrc.

**Comment by fcpg** [comment]
I tend to believe most old-timers, users and/or contributors, are on the same page, valuing stability, bugfixes in a reasonable time frame, all with a long term view.

Like @skywind3000, I would be willing to donate on a regular basis if this could help the dev efforts, one way or another. Feel free to organize something for donations when the time is right.

**Comment by Shougo** [comment]
I think github sponsor can be used for the donation.
The developers could donate part of the money makes to the ICCF if they so wish.

**Reply by skywind3000** [reply]
Consider to apply for joining some existing foundation? like Apache Foundation or something, at the right time ? 
If they can help raise money.

**Comment by skywind3000** [comment]
@mattn

Speaking of voting, there may be something we need to figure out:

- What will be voted by sponsors ?
- What will be voted by the core team (maintainers) ?
- What will be voted by others in the community ?

**Comment by gu-fan** [comment]
R.I.P Bram,  Long live the vim.

I have tested to switch to nvim for awhile a week ago, and failed and back to vim now.
I think nvim is good for newbie, but not suitable for me with my long time maintained vimrc.

**Reply by sukima** [reply]
This is me as well. I really really like my current Vim setup I've tweaked and honed over 20+ years. LUA be damned!

**Comment by sukima** [comment]
Is it possible to start a go Fund Me for friends and family. Possibly for continued maintenance of the Vim Ecosystem?

**Comment by smekkley** [comment]
I suggest closing this discussion already.  
The scope is too big and already some very details that haven't been really big issues for users are discussed, such as git tagging and even voting which are more of burdens of devs who are already busy with existing issues, PRs and Bram's todolist that yegappan has mentioned.   

People who want to discuss some topic such as how to decide about new features and donations that the existing maintainers aren't really asking for so far, should create a new more specific discussion or issue.

**Comment by slmjkdbtl** [comment]
On a side note: I hope we can freeze this repo in honor to Bram's memory and resume development on a fork. Vim without Bram and his development style will be a different software, re-organizing will be a change not unlike neovim.

**Reply by brickviking** [reply]
Personally, having been a long time vim user (since the early 2000s if not the late nineties) I'd rather see this repository continue on. I'm pretty sure Bram would have wanted that to be the case, however I'd like to see some sort of dedication within the opening screen of vim for this upcoming release at least. Disadvantages with forking is that people are going to have to go to another site than this one just to catch up with ongoing development.

That's just my personal opinion.

**Reply by slmjkdbtl** [reply]
Practically yes it might cause some confusion, but Vim has been a primary example of the "benevolent dictatorship" model, I always consider Vim Bram's thing, it's weird to me to just take over after his death and do our own thing to it and assume it's what Bram wants. Personally opinion and I know it's never going to fly, just want to say it out 🙏

**Reply by nickspoons** [reply]
It has of course always been Bram's project but he has welcomed the contributions of hundreds of other people. If he didn't want others to continue the project after his death, he wouldn't have given maintainers like Christian Brabant the repo access.

**Reply by skywind3000** [reply]
There can be a compromised solution: 

1\) freeze 9.0 to commemorate Bram
2\) move future patches in 9.1+.

**Reply by ObserverOfTime** [reply]
> There can be a compromised solution:
>
> 1) freeze 9.0 to commemorate Bram
> 2) move future patches in 9.1+.

I like this idea but I'd propose adding one more commit similar to neovim/neovim#24589 before freezing it.

**Reply by sukima** [reply]
Yes, I think a dedication to Bram in the docs it the right thing to do!

**Comment by chdiza** [comment]
I would hope that Vim continues to have a new micro version number with each patch. I see nothing wrong with this, except that I would also attach micro versions to those occasional runtime file updates.  Those micro versions are much easier to deal with than commit hashes are.

I would hope that each such patch continues to be self-contained (the full fix for the "Problem:" mentioned in the commit message), rather than as an unsquashed series of commits.

**Reply by skywind3000** [reply]
Is it possible to write a CI script for this ? once a PR is merged it will create a standalone commit and mention the previous PR's URL in its message and increase the version number by 1 ?

**Reply by ychin** [reply]
> Is it possible to write a CI script for this ? once a PR is merged it will create a standalone commit and mention the previous PR's URL in its message and increase the version number by 1 ?

I feel like this just doubles the number of commits, making the history very noisy.

**Comment by zaitmambet** [comment]
I hope Vim will continue to develop, otherwise people will just switch to NeoVim

**Reply by lokithus** [reply]
Why not merge? It's the right time and it will help both communities. Even in death, Bram helped both projects indirectly.

**Reply by Shougo** [reply]
It cannot be merged.
Because neovim has different culture.  Vim keeps backward compatibility.
But neovim can break the compatibility if it is needed.

**Reply by skywind3000** [reply]
Vim did break things as well:

1) XP support has been dropped, since 9.0.xxx
2) C89 compliance since 8.x.x ?
3) DOS support since 7.4.xxxx
4) ActiveTcl for windows daily build

**Reply by bennyyip** [reply]
Keeping backward compatibility is great. I don't think these two editor can be merged, but Vim should break backward compatibility when it is certainly necessary.
How Emacs handles backward compatibility is reasonable to me. Emacs sends deprecation warning several major versions before a feature is actually removed.

**Reply by lilydjwg** [reply]
> Vim did break things as well

But Vim never breaks things for me :-)

I tried Neovim before. It broke clipboard usage for me among other minor default option changes (that weren't documented). And from time to time I heard about Neovim users rewriting  their configuration / plugins in some more excited way (e.g. Lua).

**Reply by Shougo** [reply]
Vim may drop some old features.
But neovim's drop is not same.  It is more drastically changes.

**Reply by savchenko** [reply]
Well... "Destruction breeds creation". In my very humble opinion merge of Vim and NeoVim would be the best outcome for both.

**Reply by ychin** [reply]
While Vim does break some backwards compatibility it's usually done for *really* old platforms where continuing supporting them is a non-trivial burden. The issue at discussion here is really the culture and goals behind both projects. Neovim is a lot more aggressive on only supporting the "new" stuff.

For example, for macOS, Neovim only supports 10.15+ ([source](https://neovim.io/doc/user/support.html)). For Vim I actually don't know the minimum version but MacVim (which operates as a fork of Vim) at least supports down to 10.9, and every time I contemplate bumping MacVim's minimum requirements up there are people coming out of the woodworks saying that they rely on such old macOS versions. Are people here suggesting that these users using old OSes should just… not use updated versions of Vim anymore? Or Neovim should retroactively go back and add support for these old platforms?

Other examples include the `:hardcopy` command (used for printing files, for those who don't know). Neovim removed it (https://github.com/neovim/neovim/pull/21472) not because it's a serious burden but mostly because it doesn't support the latest features like treesitter. I think the technical details are not as important as the overall decision making process because removing a longstanding feature that people do use (even if not a huge portion of the users) just for certain ideas about what the feature should include feels like a decision that Vim would not make to me.

The APIs have also diverged quite a bit on certain features, like with how the terminals work, jobs, virtual texts etc. When people are suggesting merging, it kind of feels like they are saying that the Vim developers should just move on to Neovim, but that seems irresponsible to the existing users who rely on these features.

**Reply by savchenko** [reply]
@ychin , appreciate the points you have raised. Insightful!

> When people are suggesting merging, it kind of feels like they are saying that the Vim developers should just move on to Neovim, but that seems irresponsible to the existing users who rely on these features.

This is not something I was suggesting. Would rather see developers of both projects agreeing on a middle ground solution.

For example, `:hardcopy` is ~3500 lines of C code and still, there are many edge cases it does not handle. On the other hand, nVim supports live preview of substitution regexp  via `inccommand`, functionality that makes sense to keep in base and enabled by default.

This works in both ways.

**Reply by Shougo** [reply]
Vim developers and users are very conservative and concerned about backward compatibility.
Conversely, people who chose convenience over compatibility migrated to neovim.

**Reply by fcpg** [reply]
Requests for merging remind me of the famous "can't we all just get along".

Both projects choices and culture are significantly different, it does not make sense to merge them. They are not fiercely competing either, it is understood that if you want more cutting-edge features, you should go with NeoVim, and if you favor stability, pick Vim. This does not mean some features will never be part of Vim, but rather that the actual implementation and technical choices may differ, because things are done differently.

**Reply by ychin** [reply]
I do think that while I think both projects will exist in the forseeable future, it's useful to make sure we keep each other in mind when working on new features and designing new APIs. FWIW some of that happens today, but there isn't always a straightforward way to port Neovim changes to Vim and they are mostly done in an ad-hoc fashion. Neovim does have a wiki entry / system for porting Vim patches over though and they do that frequently.

**Reply by clason** [reply]
>  Would rather see developers of both projects agreeing on a middle ground solution.

But we are: we cooperate on features and runtime files where it makes sense (even though it might not look like it from the outside; there's even personal overlap between the core developers now); we also have long-agreed "middle ground" with vim(8)script, which both projects will continue supporting for a long time.

Anything else just doesn't make sense; "one size fits all" fits none. Most people who work on Neovim do so because they enjoy working _on Neovim_ and would never, ever contribute to Vim (again) -- and vice versa. So the assumption that "merging projects doubles the development speed" is patently false -- in fact, the opposite is true, as together we attract a wider spectrum of contributors (and contributions to one do often if not always end up in the other, at least in some form).

And let's face it, "you should join forces" is usually little more than a thinly veiled "how dare people work on stuff I don't use; they should work on what _I_ use instead!" It has little to do with any concern about the "health of the project" (whatever that means).

**Reply by savchenko** [reply]
@clason ,

Appreciate your position, clearly articulated. However, this:

> And let's face it, "you should join forces" is usually little more than a thinly veiled "how dare people work on stuff I don't use; they should work on what I use instead!" It has little to do with any concern about the "health of the project" (whatever that means).

Is blatantly false. There is much more nuance to it.

**Reply by clason** [reply]
I have yet to see this "nuance"; none of the comments I've come across (in this and other contexts; there are similar threads on Neovim and Helix "joining forces", for example) offer more than a simple drive-by-comment without details.

Happy to be proven wrong, though, if you can point me towards such a comment with a concrete, spelled out, proposition including a cost-benefit analysis?

**Reply by skywind3000** [reply]
Too many replies above. If you don’t have enough time to read through them, here’s a TL;DR version:

- Vim is similar to **Debian LTS stable**, while NeoVim is like **Ubuntu testing** without LTS.
- Not a chance to merge, and choose at your own risk.

**Reply by jgb** [reply]
I hope vim stays vim. I tried neovim, but didn't like it, it's community is too chaotic for my taste.
The amount of plugins showing up for nvim is enormous, but most of them are of low quality and are often not maintained more than a few weeks or months before everybody jumps ship to the newest shiny thing.
I don't care for LSP (it's too slow, and I just don't need it, I've got jedi for python which is perfect), treesitter is slow, buggy and makes my syntax coloring look like a tacky christmas tree and I don't like lua as a language. Also, I would miss gvim too much, I still prefer gvim over vim/nvim in a terminal.

As I said, I hope vim can stay vim, meaning a stable, steady and reliable continuation of Bram's work.
Stability, compatibility and reliability should be preferred over shiny new gimmicks.

**Reply by brickviking** [reply]
As for me, even though I'm currently working with emacs at the moment, I prefer vim on terminal, gvim has some issues with a couple of the themes I like to use. Might be something on my end, so no heat to the project for that. Old systems like mine can run vim no troubles, no idea about neovim under terminal, if it's even a thing.

TL;DR; let us each (vim/neovim) do our own thing. My uninformed opinion is "Build it. They will come." And fighting about it would not be what Bram wished for. I don't know what his opinion was regarding the neovim project, but there never seemed to be any real bad feelings. Let's keep it that way. If we try to integrate everything, it'll be a rare person who can pull it off, and most people probably won't be happy.

**Comment by Shane-XB-Qian** [comment]
someone did not know the different between `sunset` with `compatibility` ?
official vim is pursuing `compatibility`, but some out-date things (architecture or workflow etc) would be `sunset`  still.
// but `sunset` is on the judgement of 'main maintainer' (it was Bram) still, we may need to support less popular but still used ones.

recently it do changed some workflow style also, seems a bit hurry,
if PR or Change did not review and think a bit strictly or completely, the code quality maybe a bit worried.
that's a bit different like Bram did.

Anyway, we (Vimer) should keep and help Bram's legacy, and help official vim keeping its product quality.

**Reply by Shane-XB-Qian** [reply]
// this is to answer official vim vs neovim https://github.com/vim/vim/discussions/12736#discussioncomment-6759298 thread above

as my experience with Bram in repo and mail, he is very taking care `compatibility` much than `feature`, 
and follow the existed instr in doc much than some kinds of so called "no harm" changes,
and always official vim first, or work through another way to make it e.g vim9

we should help and keep that, i wish.

**Comment by skywind3000** [comment]
well, several month ago, neovim removed cscope interfaces because, so called, cscope is not maintained for a long time.

They did not extensively seek opinions from users, not a single poll, and even when many people raised objections in the issues / PR, they simply closed the conversation: https://github.com/neovim/neovim/issues/21062 https://github.com/neovim/neovim/pull/20545

I wish this should not happen in Vim.

- There are other tools still using cscope interface like: pycscope for python, starscope for ruby, and gtags-cscope for 60+ languages.
- LSP servers are not always available for certain languages or environment.
- LSP still have performance issues in large code base.
- Temporary system, for short term work, you don't want to setup complex LSP things.
- NeoVim has the motivation to drive everyone to use LSP, which is not applicable in Vim.

Plugins with job api today can simulate a cscope interface, I tried some, but none of them can provide same user experience like native cscope interface. and there are already a bunch of existing plugins, like gutentags, atags, gutentags_plus, vim-gtags, have the dependency on the native cscope interface.  Replacing it with a plugin implementation will simply break this little ecosystem.

I have been using gtags-cscope for a long time, and it has served me very well, especially in languages that lack proper language servers.

It is a tragedy that neovim removed it without any discussion.

Since Vim values backward compatibility, hope Vim can preserve the native cscope interface.

**Reply by yegappan** [reply]
As the author of the Vim9 LSP plugin (https://github.com/yegappan/lsp), I can comment on the limitations of LSP with very large code bases (hundreds of thousands of files and millions of lines of code).

LSP doesn't work properly with a very large C code base (which is typical in many large companies).  For example, in these code bases,
a symbol could be defined in multiple places to support different architectures, use cases, etc.  Also, it is typical to use '#ifdef' to handle
different architectures and platforms (similar to the Vim code base).   The LSP "Goto definition" and other features. doesn't work in these
cases.  Also, for many of the LSP functionality to work properly you need to load the files in the LSP server. When you have hundreds
of thousands of files, this simply doesn't work.

Tools like cscope, tags, GNU id-utils, GNU global and opengrok are suitable for handling such large code bases.

**Comment by SamPosh** [comment]
With the lua integration neovim's plugin eco system increased. May be vim can do similar(may be micro-python integration)

**Comment by skywind3000** [comment]
Can we renovate vim's pum window (native completion window) ?

Vim's pum window was invented 17 years ago, in version 7.0 (2006), It renders many completion systems appear a little outdated.

Here is a side-by-side comparison. The image below showcases a screenshot from nvim-cmp, which transitioned to a custom completion menu several years ago and now boasts a significantly improved and visually appealing appearance:

![图片](https://github.com/vim/vim/assets/3035071/15363976-d403-496b-aae1-9dffc65ac621)

- The popup can be adorned with a border.
- Various kind fields can be distinguished through distinct highlighting.
- Different kinds can display accompanying text alongside their icons.
- Matched characters can be emphasized through highlighting.
- Looks polished and modernized.

Another screenshot is from YCM, which is using vim's native pum window:

![264403327-872cfb7f-d5ac-45f8-a8d3-b681315e3af7](https://github.com/vim/vim/assets/3035071/92ba1271-ba14-49fe-abed-311f1d55474e)

- The kind field is restricted to display only one character, can't provide enough information.
- Can't have a border, 
- It is not possible to discern which parts of the candidates are matched.
- Looks a little out-dated.

Vim has got many great completion plugins, some offering superior responsiveness and more accurate completion results. However, when I shared my recommendation with my friends, some of them expressed that the vim could benefit from further polishing the completion window.

There are many plugins using pum window, including:

- [asyncomplete.vim](https://github.com/prabirshrestha/asyncomplete.vim)
- [ddc.vim](https://github.com/Shougo/ddc.vim)
- [vim-easycomplete](https://github.com/jayli/vim-easycomplete)
- [vim-mucomplete](https://github.com/lifepillar/vim-mucomplete)
- [vim-auto-popmenu](https://github.com/skywind3000/vim-auto-popmenu)
- [vim-lsc](https://github.com/natebosch/vim-lsc)
- [lsp](https://github.com/yegappan/lsp)
- [ycm](https://github.com/ycm-core/YouCompleteMe)

The maintainer of ycm [said](https://github.com/ycm-core/YouCompleteMe/issues/4180#issuecomment-1699677875):

> My suggestion would be to improve the Vim PUM rather than have every plugin faff around building broken ones.

This is an excellent point, and I wholeheartedly agree. If Vim can revamp the current pum window, all the plugins above can greatly benefit from it, resulting in a modernized and enhanced appearance for Vim overall.

Here are several potential improvements, listed in order of priority from high to low:

- add an optional border for pum.
- can specify a highlight group for each kind item.
- accept longer text in the kind field.
- highlight the matched text.

I believe that's enough for the next 10 years.
